File: apple/apps/ensu/Ensu/Services/ModelMaintenance.swift
import Foundation

@MainActor
protocol ModelMaintenance: AnyObject {
    func suspendMaintenance() -> ModelUseScope
    func awaitMaintenance() async
}

extension ModelMaintenance {
    func withMaintenanceSuspended<T>(_ body: () async throws -> T) async throws -> T {
        let token = suspendMaintenance()
        defer { token.close() }
        await awaitMaintenance()
        try Task.checkCancellation()
        return try await body()
    }
}

final class ModelUseScope: @unchecked Sendable {
    private let lock = NSLock()
    private var release: (@Sendable () -> Void)?
    init(_ release: @escaping @Sendable () -> Void) { self.release = release }
    func close() {
        lock.lock()
        let action = release
        release = nil
        lock.unlock()
        action?()
    }
    deinit { close() }
}

File: android/apps/ensu/app/src/main/java/io/ente/ensu/llm/ModelMaintenance.kt
package io.ente.ensu.llm

interface ModelMaintenance {
    fun suspendMaintenance(): AutoCloseable
    suspend fun awaitMaintenance()
    fun modelReadinessChanged()
}

internal suspend fun <T> ModelMaintenance?.withMaintenanceSuspended(block: suspend () -> T): T {
    val token = this?.suspendMaintenance()
    try {
        this?.awaitMaintenance()
        return block()
    } finally {
        token?.close()
    }
}
